### PR TITLE
L38: server load reporting

### PR DESCRIPTION
**Pull request by AspirinSJL**
@markdroth Please review the whole file.
@vjpai Please review the [cleanup changes](https://github.com/grpc/proposal/pull/91/files#diff-170728a03fd43428288d8b8586fad7b2R29) and the [respective rationale](https://github.com/grpc/proposal/pull/91/files#diff-170728a03fd43428288d8b8586fad7b2R37)

### CODE REVIEW

File: L30-cpp-server-load-reporting.md [hunk 1]
```diff
@@ -0,0 +1,42 @@
+﻿C++ API changes for Server Load Reporting
+----
+* Author(s): AspirinSJL
+* Approver: markdroth, vjpai
+* Status: Draft
+* Implemented in: C++
+* Last updated: 2018-07-13
+* Discussion at: TBA
```
Comment (vjpai) [inline L30-cpp-server-load-reporting.md]: Please link in a discussion. I will only add comments on the document now but we should have technical discussions on grpc-io related to this proposal.

File: L30-cpp-server-load-reporting.md [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+## Abstract
+
+As part of the load balancing functionality, the server-side load reporting service is to be introduced. A `LoadReportingServiceServerBuilderOption` will be used to enable that service. A helper function `AddLoadReportingCost()` can be used to report the customized call metrics.
```
Comment (vjpai) [inline L30-cpp-server-load-reporting.md]: What class or namespace do these new objects and functions belong to?

File: L30-cpp-server-load-reporting.md [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+A load balancer needs to get the load data of the servers in order to balance the load among the servers accordingly. One possible approach to get the load data is to request the load data from the servers via RPCs. That requires the servers to register an additional service to report the load data.
+
+
+### Related Proposals: 
+N/A
+
+## Proposal
+
+We propose a new public header `include/grpcpp/ext/server_load_reporting.h` to use the server load reporting service. Note that this public header (and the load reporting service) is only available when the binary is built with the `grpcpp_server_load_reporting` library. That library is only available in Bazel build because it depends on OpenCensus which can only be built with Bazel.
+
```
Comment (vjpai) [inline L30-cpp-server-load-reporting.md]: Will the new public header depend on anything about OpenCensus, or will it only be the implementation? If it's the header itself, we are likely to have another set of issues (related to grpc/grpc#15984, an Abseil dependence, and other issues)

File: L30-cpp-server-load-reporting.md [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+1. The header contains a `LoadReportingServiceServerBuilderOption`. The user should set that option in the `ServerBuilder` to enable the load reporting service on a server.
+2. The header also contains a helper function `void AddLoadReportingCost(grpc::ServerContext* ctx, const grpc::string& cost_name, double cost_value);`. Besides the canonical call metrics (e.g., the number of calls started, the total bytes received), the user can use this function to add other customized call metrics from their own normal services.
```
Comment (vjpai) [inline L30-cpp-server-load-reporting.md]: Why are you calling it a "helper function"? It's just a public free function, correct?

File: L30-cpp-server-load-reporting.md [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+
+We also propose two cleanup changes.
+
+1. The previous public header `include/grpc/load_reporting.h` will be renamed to `include/grpc/server_load_reporting.h` to distinguish from the client load reporting in grpclb.
+2. The previous function `void SetLoadReportingCosts(const std::vector<grpc::string>& cost_data);` in `include/grpcpp/impl/codegen/server_context.h` will be removed in favor of the new helper function.
```
Comment (vjpai) [inline L30-cpp-server-load-reporting.md]: This is apparently public API. If it is actually used anywhere, I don't think it should be removed. It was never marked experimental or anything like that.

File: L30-cpp-server-load-reporting.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+## Rationale
+
+1. The load reporting service is opted in via a `ServerBuilderOption` instead of being automatically enabled because this service should only be enabled when there will be a load balancer requesting the load data. Otherwise, the load data will be accumulated for nothing.
+2. The new API `AddLoadReportingCost()` is better than the old `SetLoadReportingCosts()` because it includes serialization step in it, which makes the API easier to use and ensures the serialization matches the deserialization in our load reporting filter.
+3. The header under `include/grpc` is renamed to have a "server_" prefix because we actually have [client load reporting in grpclb](https://github.com/grpc/grpc/blob/85daf2db65d60ebd63936a936d69c63777123d10/src/core/ext/filters/client_channel/lb_policy/grpclb/client_load_reporting_filter.h). The new name will also be more consistent with `include/grpcpp/ext/server_load_reporting.h`.
+
+## Implementation
+
+Implementation is completed. The new APIs are currently under `experimental` namespace.
```
Comment (vjpai) [inline L30-cpp-server-load-reporting.md]: Can you link to the implementation? Either as code or to the PR that created the code?

File: L30-cpp-server-load-reporting.md (hunk 3 above)
Comment (AspirinSJL) [reply L30-cpp-server-load-reporting.md]: It's only the implementation that depend on OpenCensus; the public header doesn't. So we should be safe!

File: L30-cpp-server-load-reporting.md (hunk 4 above)
Comment (AspirinSJL) [reply L30-cpp-server-load-reporting.md]: Yes, but it's not essential. The users can implement one without much effort. It only does some serialization and calls the `ServerContext::AddTrailingMetadata()` to add the entries. 

I rephrased it to "utility function" which might be clearer.

File: L30-cpp-server-load-reporting.md (hunk 5 above)
Comment (AspirinSJL) [reply L30-cpp-server-load-reporting.md]: I see. Removed this item.

File: L30-cpp-server-load-reporting.md (hunk 6 above)
Comment (AspirinSJL) [reply L30-cpp-server-load-reporting.md]: Added.

File: L30-cpp-server-load-reporting.md (hunk 2 above)
Comment (AspirinSJL) [reply L30-cpp-server-load-reporting.md]: Added.

**Comment by AspirinSJL** [comment]
Bumping this up.

**Comment by dfawley** [comment]
This cannot use L30.  Up to and including L36 is already taken.  There may be closed proposals also higher than L36.  Please include your proposal number in the PR title so it's possible for future proposals to pick a proper number.

**Comment by AspirinSJL** [comment]
@dfawley Thanks for letting me know!

@vjpai @markdroth Please take a look, thanks!

**Pull request merged**